File: app/src/main/java/org/neidhardt/dynamicsoundboard/views/edittext/CustomEditText.java
package org.neidhardt.dynamicsoundboard.views.edittext;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.util.TypedValue;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.LinearLayout;
import android.widget.TextView;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.misc.Logger;


public abstract class CustomEditText
		extends
			LinearLayout
		implements
			TextView.OnEditorActionListener,
			EditTextBackEvent.EditTextImeBackListener,
			View.OnFocusChangeListener
{
	private static final String TAG = CustomEditText.class.getName();

	EditTextBackEvent input;
	OnTextEditedListener onTextEditedListener;
	private OnFocusChangeListener onFocusChangeListener;

	public CustomEditText(Context context, AttributeSet attrs)
	{
		super(context, attrs);
		this.inflateLayout(context, this.getLayoutId());
		this.input.setOnEditorActionListener(this);
		this.input.setOnEditTextImeBackListener(this);
		this.input.setOnFocusChangeListener(this);

		TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.CustomEditText, 0, 0);
		float size = array.getDimension(R.styleable.CustomEditText_text_size, 0);
		if (array.hasValue(R.styleable.CustomEditText_text_size))
			this.input.setTextSize(TypedValue.COMPLEX_UNIT_PX, size);

		int color = array.getColor(R.styleable.CustomEditText_text_color, 0);
		if (array.hasValue(R.styleable.CustomEditText_text_color))
			this.input.setTextColor(color);

		array.recycle();
	}

	public void setOnTextEditedListener(OnTextEditedListener listener)
	{
		this.onTextEditedListener = listener;
	}

	@SuppressWarnings("unused")
	public void addTextChangedListener(TextWatcher listener)
	{
		this.input.addTextChangedListener(listener);
	}

	@Override
	public boolean hasFocus()
	{
		return this.input.hasFocus();
	}

	@Override
	public void clearFocus()
	{
		this.input.clearFocus();
	}

	@Override
	public void setOnFocusChangeListener(OnFocusChangeListener onFocusChangeListener)
	{
		this.onFocusChangeListener = onFocusChangeListener;
	}

	@Override
	public boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent)
	{
		if (actionId == EditorInfo.IME_ACTION_DONE)
		{
			if (this.onTextEditedListener != null)
				this.onTextEditedListener.onTextEdited(this.input.getText().toString());
		}
		return false;
	}

	@Override
	public void onImeBack(EditTextBackEvent ctrl, String text)
	{
		if (this.onTextEditedListener != null)
			this.onTextEditedListener.onTextEdited(this.input.getText().toString());
	}

	@Override
	public void onFocusChange(View v, boolean hasFocus)
	{
		if (this.onTextEditedListener != null && !hasFocus)
			this.onTextEditedListener.onTextEdited(this.input.getText().toString());
		if (this.onFocusChangeListener != null)
			this.onFocusChangeListener.onFocusChange(this, hasFocus);
	}

	protected abstract int getLayoutId();

	void inflateLayout(Context context, int layoutToInflate)
	{
		LayoutInflater.from(context).inflate(layoutToInflate, this, true);
		this.input = (EditTextBackEvent) this.findViewById(R.id.edittext);
	}

	public void setText(String text)
	{
		if (this.input != null)
			this.input.setText(text);
	}

	public Editable getText()
	{
		if (this.input != null)
			return this.input.getText();
		return null;
	}

	public String getDisplayedText()
	{
		String userInput = this.getText().toString();
		if (!userInput.isEmpty())
			return userInput;
		return this.getHint().toString();
	}

	@SuppressWarnings("unused")
	public void setHint(String hint)
	{
		if (this.input != null)
			this.input.setHint(hint);
	}

	CharSequence getHint()
	{
		if (this.input != null)
			return this.input.getHint();
		return null;
	}

	@NonNull
	@Override
	protected Parcelable onSaveInstanceState()
	{
		Logger.d(TAG, "onSaveInstanceState");
		Parcelable superState = super.onSaveInstanceState();
		return new SavedCustomEditTextState(superState, this.input.getText().toString());
	}

	@Override
	protected void onRestoreInstanceState(Parcelable state)
	{
		Logger.d(TAG, "onRestoreInstanceState " + state);
		SavedCustomEditTextState savedState = (SavedCustomEditTextState) state;
		super.onRestoreInstanceState(savedState.getSuperState());

		if (savedState.getValue() != null)
			this.input.setText(savedState.getValue());
	}

	@SuppressWarnings("unchecked")
	@Override
	protected void dispatchSaveInstanceState(@NonNull SparseArray container)
	{
		// As we save our own instance state, ensure our children don't save
		// and restore their state as well.
		super.dispatchFreezeSelfOnly(container);
	}

	@SuppressWarnings("unchecked")
	@Override
	protected void dispatchRestoreInstanceState(@NonNull SparseArray container)
	{
		/** See comment in {@link #dispatchSaveInstanceState(android.util.SparseArray)} */
		super.dispatchThawSelfOnly(container);
	}

	public static class SavedCustomEditTextState extends BaseSavedState
	{
		private String value;

		public String getValue()
		{
			return this.value;
		}

		public SavedCustomEditTextState(Parcelable superState, String value)
		{
			super(superState);
			this.value = value;
		}

		private SavedCustomEditTextState(Parcel superState)
		{
			super(superState);
			this.value = superState.readString();
		}

		@Override
		public void writeToParcel(@NonNull Parcel destination, int flags)
		{
			super.writeToParcel(destination, flags);
			destination.writeString(this.value);
		}

		public static final Parcelable.Creator<SavedCustomEditTextState> CREATOR = new
				Parcelable.Creator<SavedCustomEditTextState>()
				{
					public SavedCustomEditTextState createFromParcel(@NonNull Parcel in)
					{
						return new SavedCustomEditTextState(in);
					}

					@NonNull
					public SavedCustomEditTextState[] newArray(int size)
					{
						return new SavedCustomEditTextState[size];
					}
				};
	}

	public interface OnTextEditedListener
	{
		void onTextEdited(String text);
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/views/edittext/CustomEditText.kt
package org.neidhardt.dynamicsoundboard.views.edittext

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.text.SpannableStringBuilder
import android.util.AttributeSet
import android.util.SparseArray
import android.util.TypedValue
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.FrameLayout
import android.widget.TextView
import org.neidhardt.dynamicsoundboard.R

abstract class CustomEditText :
		FrameLayout,
		TextView.OnEditorActionListener,
		EditTextBackEvent.EditTextImeBackListener,
		View.OnFocusChangeListener
{
	abstract var input: EditTextBackEvent?
	abstract var onTextEditedListener: OnTextEditedListener?

	var text: String?
		get() { return this.input!!.text.toString() }
		set(value) { this.input!!.text = SpannableStringBuilder(value) }

	val displayedText: String
		get()
		{
			val userInput = this.text!!.toString()
			if (!userInput.isEmpty())
				return userInput
			return this.hint!!.toString()
		}

	var hint: CharSequence?
		get()
		{
			return this.input!!.hint
		}
		set(value)
		{
			this.input!!.hint = value
		}

	constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
	{
		this.inflateLayout(context)
		this.initTextListeners()
		this.setValuesAttributeSet(context, attrs)
	}

	abstract fun inflateLayout(context: Context)

	fun initTextListeners()
	{
		this.input!!.setOnEditorActionListener(this)
		this.input!!.setOnEditTextImeBackListener(this)
		this.input!!.onFocusChangeListener = this
	}

	fun setValuesAttributeSet(context: Context, attrs: AttributeSet)
	{
		val array = context.obtainStyledAttributes(attrs, R.styleable.CustomEditText, 0, 0)
		val size = array.getDimension(R.styleable.CustomEditText_text_size, 0f)
		if (array.hasValue(R.styleable.CustomEditText_text_size))
			this.input!!.setTextSize(TypedValue.COMPLEX_UNIT_PX, size)

		val color = array.getColor(R.styleable.CustomEditText_text_color, 0)
		if (array.hasValue(R.styleable.CustomEditText_text_color))
			this.input!!.setTextColor(color)

		array.recycle()
	}

	override fun hasFocus(): Boolean
	{
		return this.input!!.hasFocus()
	}

	override fun clearFocus()
	{
		this.input!!.clearFocus()
	}

	override fun onEditorAction(textView: TextView, actionId: Int, keyEvent: KeyEvent): Boolean
	{
		if (actionId == EditorInfo.IME_ACTION_DONE)
		{
			this.onTextEditedListener?.onTextEdited(this.input!!.text.toString())
		}
		return false
	}

	override fun onImeBack(ctrl: EditTextBackEvent, text: String)
	{
		this.onTextEditedListener?.onTextEdited(this.input!!.text.toString())
	}

	override fun onFocusChange(v: View, hasFocus: Boolean)
	{
		if (!hasFocus)
			this.onTextEditedListener?.onTextEdited(this.input!!.text.toString())
			this.onFocusChangeListener?.onFocusChange(this, hasFocus)
	}

	override fun onSaveInstanceState(): Parcelable
	{
		val superState = super.onSaveInstanceState()
		return SavedCustomEditTextState(superState, this.input!!.text.toString())
	}

	override fun onRestoreInstanceState(state: Parcelable?)
	{
		val savedState = state as SavedCustomEditTextState?
		super.onRestoreInstanceState(savedState?.superState)

		if (savedState?.value != null)
			this.input!!.setText(savedState?.value)
	}

	@SuppressWarnings("unchecked")
	protected override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>?)
	{
		// As we save our own instance state, ensure our children don't save
		// and restore their state as well.
		super.dispatchFreezeSelfOnly(container)
	}

	@SuppressWarnings("unchecked")
	protected override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>?) {
		/** See comment in [.dispatchSaveInstanceState]  */
		super.dispatchThawSelfOnly(container)
	}

	class SavedCustomEditTextState : View.BaseSavedState
	{
		var value: String? = null
			private set

		constructor(superState: Parcelable, value: String) : super(superState)
		{
			this.value = value
		}

		private constructor(superState: Parcel) : super(superState)
		{
			this.value = superState.readString()
		}

		override fun writeToParcel(destination: Parcel, flags: Int)
		{
			super.writeToParcel(destination, flags)
			destination.writeString(this.value)
		}

		companion object
		{
			val CREATOR: Parcelable.Creator<SavedCustomEditTextState> = object : Parcelable.Creator<SavedCustomEditTextState>
			{
				override fun createFromParcel(`in`: Parcel): SavedCustomEditTextState {
					return SavedCustomEditTextState(`in`)
				}

				override fun newArray(size: Int): Array<SavedCustomEditTextState?> {
					return arrayOfNulls(size)
				}
			}
		}
	}

	interface OnTextEditedListener
	{
		fun onTextEdited(text: String)
	}

}
